pep-9999.rst

```diff
@@ -0,0 +1,267 @@
+PEP: 9999
+Title: Enriching Exceptions with Notes
+Author: Zac Hatfield-Dodds <zac@zhd.dev>
+Sponsor: Irit Katriel
+Status: Draft
+Type: Standards Track
+Requires: 654
+Created: 20-Dec-2021
+Python-Version: 3.11
+Post-History:
+
+
+Abstract
+========
+Exception objects are typically initialized with a message that describes the
+error which has occured.  Because further information may be available when the
+exception is caught and re-raised, this PEP proposes to add a ``.__note__``
+attribute and update the builtin traceback formatting code to include it in
+the formatted traceback following the exception string.
+
+This is particularly useful in relation to :pep:`654` ``ExceptionGroup``s, which
+make previous workarounds ineffective or confusing.  Use-cases have been identified
+in the standard library, Hypothesis package, and common code patterns with retries.
+
+
+Motivation
+==========
+When an exception is created in order to be raised, it is usually initialized
+with information that describes the error that has occurred. There are cases
+where it is useful to add information after the exception was caught.
+For example,
+
+- testing libraries may wish to show the values involved in a failing assertion,
+  or the steps to reproduce a failure (e.g. `pytest` and `hypothesis`; example below).
+- code with retries may wish to note which iteration or timestamp raised which
+  error - especially if re-raising them in an ``ExceptionGroup``
+- programming environments for novices can provide more detailed descriptions
+  of various errors, and tips for resolving them (e.g. `friendly-traceback`).
+
+Existing approaches must pass this additional information around while keeping
+it in-sync with the state of raised, and potentially caught or chained, exceptions.
+This is already error-prone, and made more difficult by :pep:`654` ``ExceptionGroup``s,
+so the time is right for a built-in solution.  We therefore propose to add a mutable
+field ``__note__`` to ``BaseException``, which can be assigned a string - and
+if assigned, is automatically displayed in formatted tracebacks.
+
+
+Example usage
+-------------
+
+   >>> try:
+   ...     raise TypeError('bad type')
+   ... except Exception as e:
+   ...     e.__note__ = 'Add some information'
+   ...     raise
+   ...
+   Traceback (most recent call last):
+     File "<stdin>", line 2, in <module>
+   TypeError: bad type
+   Add some information
+   >>>
+
+When collecting exceptions into an exception group, we may want
+to add context information for the individual errors. In the following
+example with `Hypothesis' proposed support for ExceptionGroup
+<https://github.com/HypothesisWorks/hypothesis/pull/3191>`__, each
```

Comment:
```suggestion
<https://github.com/HypothesisWorks/hypothesis/pull/3191>`_, each
```

Reply:
See https://stackoverflow.com/a/27435441/9297601 - two underscores avoids defining a link target.  Doesn't make much difference here, but I'm in the habit 😅